plugins/modeltraining/pkg/modeltraining/aggregation.go

```diff
@@ -0,0 +1,169 @@
+	Buckets  []Bucket `json:"buckets"`
+}
+
+type Bucket struct {
+	Key      Key `json:"key"`
+	DocCount int `json:"doc_count"`
+}
+
+type Aggregations struct {
+	ByCluster map[string]*ClusterAggregation `json:",inline,omitEmpty"`
+}
+
+type ClusterAggregation struct {
+	ByNamespace map[string]*NamespaceAggregation `json:",inline,omitEmpty"`
+}
+
+type NamespaceAggregation struct {
+	ByDeployment map[string]*DeploymentLogCount `json:",inline,omitEmpty"`
+}
+
+type DeploymentLogCount struct {
+	DeploymentName string `json:"deployment_name"`
+	Count          int    `json:"doc_count"`
+}
+
+func newAggregations() *Aggregations {
+	return &Aggregations{ByCluster: make(map[string]*ClusterAggregation)}
+}
+
+func (a *Aggregations) Add(bucket Bucket) {
+	if _, ok := a.ByCluster[bucket.Key.ClusterID]; !ok {
+		a.ByCluster[bucket.Key.ClusterID] = &ClusterAggregation{
+			ByNamespace: make(map[string]*NamespaceAggregation),
+		}
+	}
+	cluster := a.ByCluster[bucket.Key.ClusterID]
+	if _, ok := cluster.ByNamespace[bucket.Key.NamespaceName]; !ok {
+		cluster.ByNamespace[bucket.Key.NamespaceName] = &NamespaceAggregation{
+			ByDeployment: make(map[string]*DeploymentLogCount),
+		}
+	}
+	a.ByCluster[bucket.Key.ClusterID].ByNamespace[bucket.Key.NamespaceName].ByDeployment[bucket.Key.DeploymentName] = &DeploymentLogCount{
+		DeploymentName: bucket.Key.DeploymentName,
+		Count:          bucket.DocCount,
+	}
+}
+
+func (s *ModelTrainingPlugin) run_aggregation() {
+	for {
+		request := map[string]interface{}{
+			"size": 0,
+			"query": map[string]interface{}{
+				"bool": map[string]interface{}{
+					"must": []map[string]interface{}{
+						{
+							"match": map[string]interface{}{
+								"log_type": "workload",
+							},
+						},
+						{
+							"regexp": map[string]interface{}{
+								"kubernetes.namespace_name.keyword": ".+",
+							},
+						},
+						{
+							"regexp": map[string]interface{}{
+								"deployment.keyword": ".+",
+							},
+						},
+					},
+				},
+			},
+			"aggs": map[string]interface{}{
+				"bucket": map[string]interface{}{
+					"composite": map[string]interface{}{
+						"size": 4,
+						"sources": []map[string]interface{}{
+							{
+								"cluster_id": map[string]interface{}{
+									"terms": map[string]interface{}{
+										"field": "cluster_id",
+									},
+								},
+							},
+							{
+								"namespace_name": map[string]interface{}{
+									"terms": map[string]interface{}{
+										"field": "kubernetes.namespace_name.keyword",
+									},
+								},
+							},
+							{
+								"deployment_name": map[string]interface{}{
+									"terms": map[string]interface{}{
+										"field": "deployment.keyword",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		resultAgg := newAggregations()
+		for {
+			var buf bytes.Buffer
+			//var r map[string]interface{}
+			if err := json.NewEncoder(&buf).Encode(request); err != nil {
+				log.Fatalf("Error getting response: %s", err)
```

Comment:
This can be a fatal error (since it would be a bug) but the error message doesn't match.